`warn_no_constructor_for_refconst` can not be disabled (Boost.ConceptCheck)

|  |  |
| --- | --- |
| Bugzilla Link | [41759](https://llvm.org/bz41759) |
| Version | unspecified |
| OS | Linux |
| CC | @dwblaikie,@DougGregor,@mizvekov,@zygoloid |

## Extended Description 
The `warn_no_constructor_for_refconst` (and possibly other from close category) can not be disabled as there is no corresponding `-Wno` flag available  in clang++<=9.

This causes issues while compiling code that makes use of Boost.ConceptCheck library. For example:


```cpp
#include <boost/concept_check.hpp>
int main()
{
    using T = int const&;
    boost::function_requires
        <
            boost::EqualityComparableConcept<T>
        >();
}

```

will trigger a flood of warnings:


```console
$ clang++ -std=c++11 -I/mnt/d/boost.wsl clang_warn_no_constructor_for_refconst.cpp
In file included from clang_warn_no_constructor_for_refconst.cpp:1:
/mnt/d/boost.wsl/boost/concept_check.hpp:231:17: warning: struct 'EqualityComparable<const int &>' does not declare any constructor
      to initialize its non-modifiable members
  BOOST_concept(EqualityComparable,(TT))
                ^
/mnt/d/boost.wsl/boost/concept_check.hpp:231:17: note: in instantiation of template class 'boost::EqualityComparable<const int &>'
      requested here
/mnt/d/boost.wsl/boost/concept/detail/has_constraints.hpp:32:63: note: in instantiation of template class
      'boost::EqualityComparableConcept<const int &>' requested here
  inline yes has_constraints_(Model*, wrap_constraints<Model,&Model::constraints>* = 0);
                                                              ^
/mnt/d/boost.wsl/boost/concept/detail/has_constraints.hpp:44:25: note: while substituting deduced template arguments into function
      template 'has_constraints_' [with Model = boost::EqualityComparableConcept<const int &>]
      , value = sizeof( detail::has_constraints_((Model*)0) ) == sizeof(detail::yes) );
                        ^
/mnt/d/boost.wsl/boost/concept/detail/general.hpp:53:19: note: in instantiation of template class
      'boost::concepts::not_satisfied<boost::EqualityComparableConcept<const int &> >' requested here
        concepts::not_satisfied<Model>::value
                  ^
/mnt/d/boost.wsl/boost/concept_check.hpp:50:7: note: in instantiation of template class 'boost::concepts::requirement_<void
      (*)(boost::EqualityComparableConcept<const int &>)>' requested here
      BOOST_CONCEPT_ASSERT((Model));
      ^
/mnt/d/boost.wsl/boost/concept/assert.hpp:43:5: note: expanded from macro 'BOOST_CONCEPT_ASSERT'
    BOOST_CONCEPT_ASSERT_FN(void(*)ModelInParens)
    ^
/mnt/d/boost.wsl/boost/concept/detail/general.hpp:71:25: note: expanded from macro 'BOOST_CONCEPT_ASSERT_FN'
    &::boost::concepts::requirement_<ModelFnPtr>::failed>    \
                        ^
clang_warn_no_constructor_for_refconst.cpp:6:12: note: in instantiation of function template specialization
      'boost::function_requires<boost::EqualityComparableConcept<const int &> >' requested here
    boost::function_requires
           ^
/mnt/d/boost.wsl/boost/concept_check.hpp:238:8: note: reference member 'a' will never be initialized
    TT a, b;
       ^
/mnt/d/boost.wsl/boost/concept_check.hpp:238:11: note: reference member 'b' will never be initialized
    TT a, b;
          ^
1 warning generated.
```


There is no way to disable this

```
warning: struct 'EqualityComparable<const int &>' does not declare any constructor to initialize its non-modifiable members
```

Although I understand and agree with rationale provided by Chandler here https://softwareengineering.stackexchange.com/questions/122608/clang-warning-flags-for-objective-c-development/124574#124574, to warning about "never really correct" code, I think that in this particular use case of the Boost.ConceptCheck library, it is Okey to expect to be able to disable the `warn_no_constructor_for_refconst` (and similar).

The PITA is that for any non-trivial size of templated library that makes use of  Boost.ConceptCheck, compilation with Clang leads to a flood of warning.
As side effect, it causes trouble when compiling on popular CI services available for open source projects (timeouts, builds killed due to logs too long  etc.)

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -385,6 +385,10 @@ Improvements to Clang's diagnostics
 
   Fixes #GH131127
 
+- ``-Wuninitialized`` now diagnoses when a class does not declare any
+  constructors to initialize their non-modifiable members. The diagnostic is
+  not new; being controlled via a warning group is what's new. Fixes #GH41104
+
 Improvements to Clang's time-trace
 ----------------------------------
 

--- a/clang/include/clang/Basic/DiagnosticSemaKinds.td
+++ b/clang/include/clang/Basic/DiagnosticSemaKinds.td
@@ -2266,7 +2266,8 @@ def err_constructor_byvalue_arg : Error<
   "copy constructor must pass its first argument by reference">;
 def warn_no_constructor_for_refconst : Warning<
   "%select{struct|interface|union|class|enum}0 %1 does not declare any "
-  "constructor to initialize its non-modifiable members">;
+  "constructor to initialize its non-modifiable members">,
+  InGroup<Uninitialized>;
 def note_refconst_member_not_initialized : Note<
   "%select{const|reference}0 member %1 will never be initialized">;
 def ext_ms_explicit_constructor_call : ExtWarn<

--- a/clang/test/Misc/warning-flags.c
+++ b/clang/test/Misc/warning-flags.c
@@ -18,7 +18,7 @@ This test serves two purposes:
 
 The list of warnings below should NEVER grow.  It should gradually shrink to 0.
 
-CHECK: Warnings without flags (57):
+CHECK: Warnings without flags (56):
 
 CHECK-NEXT:   ext_expected_semi_decl_list
 CHECK-NEXT:   ext_missing_whitespace_after_macro_name
@@ -57,7 +57,6 @@ CHECK-NEXT:   warn_method_param_redefinition
 CHECK-NEXT:   warn_missing_case_for_condition
 CHECK-NEXT:   warn_missing_dependent_template_keyword
 CHECK-NEXT:   warn_missing_whitespace_after_macro_name
-CHECK-NEXT:   warn_no_constructor_for_refconst
 CHECK-NEXT:   warn_not_compound_assign
 CHECK-NEXT:   warn_objc_property_copy_missing_on_block
 CHECK-NEXT:   warn_objc_protocol_qualifier_missing_id

--- a/clang/test/SemaCXX/uninitialized-no-ctor.cpp
+++ b/clang/test/SemaCXX/uninitialized-no-ctor.cpp
@@ -1,0 +1,13 @@
+// RUN: %clang_cc1 -fsyntax-only -Wuninitialized -verify %s
+// RUN: %clang_cc1 -fsyntax-only -Wno-uninitialized  -verify=good %s
+//good-no-diagnostics
+
+template <class T>
+class RefMem { // expected-warning {{class 'RefMem<int &>' does not declare any constructor to initialize its non-modifiable members}}
+  T &M; // expected-note {{reference member 'M' will never be initialized}}
+};
+
+struct RefRef {
+  RefMem<int &> R; // expected-note {{in instantiation of template class 'RefMem<int &>' requested here}}
+};
+

